Login improvements

BudsieBuds opened the pull request:
See #3190 for more info.

molomixx commented:
Ok

jekkos commented:
@BudsieBuds is the other PR abandoned in favor of this one now?

jekkos commented on Gruntfile.js:
```diff
@@ -58,10 +49,89 @@ module.exports = function(grunt) {
+		copy: {
+			themes: {
+				files: [{
+					expand: true,
```
indentation here

BudsieBuds commented:
@jekkos Sure, this one was added because I created a new branch. The other one was from the forks master. In hindsight, that wasn't ideal in the first place.

jekkos commented:
Ok should have known that, indeed. I'll give it a last look tonight. Also some testing/feedback from other contributors is welcomed @opensourcepos/org

jekkos commented:
Will also have a look at this codeQL thingy.. looks cool.

BudsieBuds commented:
Didn't mean to implement it, that was also a disadvantage of working in the master hehe. But yeah, don't know if I set it up in the optimal way. I think the GitHub actions we should look at are super-linter and PHP composer workflows.

There are a lot of bugs on the dev server right now, but don't know what does and what doesn't have to do with my PR. Most screens don't show normal content, glyphicons seem broken and switching a tab on the config screen sets the user back to home.

jekkos commented:
It looks like there is an issue in the minification. Maybe this is because of the new bootstrap libs, they might contain code that the grunt minifier is not able to handle.

![Screenshot from 2021-06-15 15-54-32](https://user-images.githubusercontent.com/576234/122065151-fff52380-cdf1-11eb-87bc-bba20b03f78f.png)

jekkos commented:
what I did last time to fix this is exclude some libs from the minification, and include the minified versions directly.

jekkos commented:
@BudsieBuds if you add `?debug=true` you will load a version w/o minified JS. This one seems to work.

BudsieBuds commented:
@jekkos Yes, the config works with the `?debug=true` tag. Every theme seems to work fine, except for the Docs theme. It doesn't load the correct CSS file.

I didn't load the BS5 JS, because it's not needed for the login screen. I think the problem is in the way I changed the handling of the BS3 JS file.
Let me think out loud here: I've added it to the Gruntfile to copy it to the `public/dist/bootstrap/js/` and `grunt-tags` adds it to the `header.php`. The final difference is the following:
```diff
<!-- bower:js -->
...
- <script src="bower_components/bootstrap/dist/js/bootstrap.js"></script>
...
<!-- endbower -->
<!-- start js template tags -->
...
+ <script type="text/javascript" src="dist/bootstrap/js/bootstrap.min.js"></script>
...
<!-- end js template tags -->
```

Could something go wrong handling this?

I diffed the `opensourcepos.min.js` from the demo and the dev. There are some small differences, but the major diff is that the BS JS seems to be missing. It has this instead of the full file:
```
Package.describe({
  name: "twbs:bootstrap",
  summary:
    "The most popular front-end framework for developing responsive, mobile first projects on the web.",
  version: "5.0.1",
  git: "https://github.com/twbs/bootstrap.git",
}),
Package.onUse((t) => {
  t.versionsFrom("METEOR@1.0"),
    t.addFiles(["dist/css/bootstrap.css", "dist/js/bootstrap.js"], "client");
})(
```

Any suggestions?

jekkos commented:
> @jekkos Yes, the config works with the `?debug=true` tag. Every theme seems to work fine, except for the Docs theme. It doesn't load the correct CSS file.
> 
> I didn't load the BS5 JS, because it's not needed for the login screen. I think the problem is in the way I changed the handling of the BS3 JS file.
> Let me think out loud here: I've added it to the Gruntfile to copy it to the `public/dist/bootstrap/js/` and `grunt-tags` adds it to the `header.php`. The final difference is the following:
> 
> ```diff
> <!-- bower:js -->
> ...
> - <script src="bower_components/bootstrap/dist/js/bootstrap.js"></script>
> ...
> <!-- endbower -->
> <!-- start js template tags -->
> ...
> + <script type="text/javascript" src="dist/bootstrap/js/bootstrap.min.js"></script>
> ...
> <!-- end js template tags -->
> ```
> 
> Could something go wrong handling this?
> 
> I diffed the `opensourcepos.min.js` from the demo and the dev. There are some small differences, but the major diff is that the BS JS seems to be missing. It has this instead of the full file:
> 
> ```
> Package.describe({
>   name: "twbs:bootstrap",
>   summary:
>     "The most popular front-end framework for developing responsive, mobile first projects on the web.",
>   version: "5.0.1",
>   git: "https://github.com/twbs/bootstrap.git",
> }),
> Package.onUse((t) => {
>   t.versionsFrom("METEOR@1.0"),
>     t.addFiles(["dist/css/bootstrap.css", "dist/js/bootstrap.js"], "client");
> })(
> ```
> 
> Any suggestions?
Thanks for the analysis.

That last code snippet seems to be the issue. It was not included before I think and it contains some statements that we can't minify properly. So in some cases it's better to include the minified JS directly instead of adding the original source and try to minify it ourselves (due to some specific statements that might crash the minification).

jekkos commented on Gruntfile.js:
```diff
@@ -145,7 +215,7 @@ module.exports = function(grunt) {
 					ignorePath: '../../../public/'
 				},
-				src: ['public/js/jquery*', 'public/js/*.js'],
+				src: ['public/dist/bootstrap/js/*.min.js', 'public/js/jquery*', 'public/js/*.js'],
```
@BudsieBuds guessing that the issue might be here. Can you remove the bootstrap from the minification? Let's try to include it directly instead

jekkos commented on Gruntfile.js:
```diff
@@ -269,7 +341,7 @@ module.exports = function(grunt) {
 	grunt.loadNpmTasks('grunt-contrib-compress');
 
-	grunt.registerTask('default', ['wiredep', 'bower_concat', 'bowercopy', 'concat', 'uglify', 'cssmin', 'tags', 'cachebreaker']);
+	grunt.registerTask('default', ['wiredep', 'bower_concat', 'bowercopy', 'copy', 'concat', 'uglify', 'cssmin', 'tags', 'cachebreaker']);
```
I think the issue is that the copy happens before the concat. The concat will pick up this bootstrap source file and append it to the file that will be minified. Maybe we can switch order here or make sure that the bootstrap original source file is not included in the concat.

jekkos replied:
Maybe let's try to remove the bootstrap source file first from the grunt build to see if that resolves the issue

jekkos replied:
@BudsieBuds If we use files from npm that are already minified, we can just copy them after the others are minified. So then we are certain that Grunt won't pick them up (which is the issue here apparantly).

jekkos commented on Gruntfile.js:
```diff
@@ -58,10 +49,89 @@ module.exports = function(grunt) {
+					{
+					expand: true,
+					cwd: 'node_modules/bootstrap/dist/js',
+					src: ['bootstrap.js', 'bootstrap.min.js'],
```
What you could try as a first thing is remove the original source file here, so if we concat something, we only concat the minified file.

jekkos replied:
Could you try to move copy step in Grunt after uglify?

jekkos replied:
If you want then I can have a look here.

BudsieBuds commented:
@jekkos It passed the checks. I'll wait until the dev server is updated to test it out.

jekkos commented:
@BudsieBuds I tested locally and this fix works, so moving the copy after the uglify.

jekkos commented:
I have pushed the fix.. let's wait for the build.

BudsieBuds commented:
Should I revert the changes I made in commit a7e2b8f?

jekkos commented:
It seems to be broken again indeed.
Locally I added my commit on top of the two original ones, so after the `Create codeql-analysis.yml`

I did a git reset --hard to that commit and will rebuild now to make sure it's fine. If so then I can force push to this branch if you want.

jekkos commented:
Ok seems fine on my local setup with that fix on top of your two commits. Let's start from there.

BudsieBuds commented:
> It seems to be broken again indeed.
> Locally I added my commit on top of the two original ones, so after the `Create codeql-analysis.yml`
> 
> I did a git reset --hard to that commit and will rebuild now to make sure it's fine. If so then I can force push to this branch if you want.

Yes, please do. If it's working I've got some small changes queud and after that I guess it's ready to be merged to master.

jekkos commented:
Ok last version deployed still has this issue, I'll have to take another look, maybe somewhere this weekend.

BudsieBuds commented:
@jekkos The issue with the glyphyicons fonts seems fixed, but the JS issue remains.

BudsieBuds commented:
Retrying with grunt back to version 0.4.5. It was updated in the package.json for some reason.

jekkos commented:
@BudsieBuds I don't understand why we have this isssue. The container on dev is the exact last commit you pushed here. Locally it works fine. I run the build with the same command as the one travis is using..

jekkos commented:
Am trying to remove node_modules and dist/bower_components locally to do a rebuild. Maybe it's related to the buildbox upgrade.

jekkos commented:
Indeed after this refresh I can now reproduce the issue locally.

jekkos commented:
Travis currently uses npm 6.14.12, grunt 1.4.0
we might want to try and install grunt-cli using npm instead

jekkos commented:
The culprit is bower, for some reason it's download bootstrap 5 now and it adds it to the files to be minified. This doesn't work well as it contains some newer ES6 syntax. 
As the idea was to get rid of boostrap via bower, I added an exclude for bootstrap and use the copy file in the concat step.

This however creates a new issue, as bootstrap itself depends on jQuery. I'd say that in this case we should include jQuery as the very first dependency, and to do this we might need to migrate it to npm as well.

jekkos commented:
@BudsieBuds I pushed my WIP on this.

jekkos commented:
So the issue here mainly is that once a lib is moved to npm, we should also find the proper order in which to include files  for the minifcation. 

The concat step will add bower managed dependencies in the correct order as it sees how the dependencies are related.
An alternative to moving jquery to npm would be to add back the bootswatch and only copy bs5 using npm for now.

BudsieBuds commented:
@jekkos Any solution is fine with me at this point. In the future, we can take a look at how to optimize everything. I think the main thing now is to just get it working.

jekkos commented:
@BudsieBuds I had to revert to option2. Keep the grunt build more or less as is and use npm to download bootstrap 5 for the login page only. So locally the javascript works fine. Will check dev server later on.

jekkos commented:
Ok I had to revert one more thing in the buildbox, it's on dev now and it looks fine.

jekkos commented:
I think we are good for this one. Any objections before merge?

BudsieBuds commented:
@jekkos Very good! I did encounter a bug, try submitting a change in the config screen. It gives the success message in a new screen without markup. Also, the language files should be updated for other languages. Can you do this automagically with Weblate?

jekkos commented:
@BudsieBuds I had a post receive hook in weblate that would sync these strings automatically. But it might be broken so I'll do it manually later this week.

jekkos commented:
@BudsieBuds I have rebased this branch with the latest fix on master and the config save works fine on my end.

jekkos commented:
Language files should also be synced now.

jekkos commented:
Deployed on dev and looks fine at first glance, would be good if some more folks could do some sanity checks before the merge? @opensourcepos/org

BudsieBuds commented:
I've played around with it a little, tested out all the themes. Everything seems fine to me.
